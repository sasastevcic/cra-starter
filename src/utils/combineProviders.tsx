/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (providers: Providers): FunctionComponent<unknown> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
