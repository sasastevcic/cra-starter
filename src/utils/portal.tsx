import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
