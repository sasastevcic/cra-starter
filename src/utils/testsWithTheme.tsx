/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
