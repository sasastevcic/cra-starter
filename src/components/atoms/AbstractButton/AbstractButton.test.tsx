import renderer from 'react-test-renderer';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import AbstractButton from './index';

describe('AbstractButton', () => {
	it('should render', () => {
		const { getByTestId } = renderWithTheme(<AbstractButton />);

		const element = getByTestId('AbstractButton');

		expect(element).toBeInTheDocument();
	});

	it('should render anchor tag with href as prop', () => {
		const { getByTestId } = renderWithTheme(<AbstractButton href="https://www.google.com" />);

		const element = getByTestId('AbstractButton');

		expect(element).toContainHTML('a');
	});

	it('should render button tag as default', () => {
		const { getByTestId } = renderWithTheme(<AbstractButton />);

		const element = getByTestId('AbstractButton');

		expect(element).toContainHTML('button');
	});

	it('should render button correctly', () => {
		const tree = renderer.create(<AbstractButton />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render submit button correctly', () => {
		const tree = renderer.create(<AbstractButton type="submit" />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render anchor correctly', () => {
		const tree = renderer.create(<AbstractButton href="https://www.google.com" />).toJSON();

		expect(tree).toMatchSnapshot();
	});
});
