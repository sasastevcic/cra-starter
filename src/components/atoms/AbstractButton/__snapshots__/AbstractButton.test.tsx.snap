// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractButton should render anchor correctly 1`] = `
<a
  data-testid="AbstractButton"
  href="https://www.google.com"
  rel="noopener noreferrer"
  target="_blank"
/>
`;

exports[`AbstractButton should render button correctly 1`] = `
<button
  data-testid="AbstractButton"
  type="button"
/>
`;

exports[`AbstractButton should render submit button correctly 1`] = `
<button
  data-testid="AbstractButton"
  type="submit"
/>
`;
